posts/2018-10-06-intro-to-mtl.md

```diff
@@ -0,0 +1,491 @@
+
+### Effect type classes
+
+`ApplicativeAsk` is an example for what is at the core of Cats-mtl.
+Cats-mtl provides type classes for most common effects which let you choose what kind of effects you need without commiting to a specific monad transformer stack.
+
+Ideally, you'd write all your code using only an abstract type constructor `F[_]` with different type class constraints and then at the end run that code with a specific data type that is able to fulfill those constraints.
+
+So without further ado, let's try to convert our `Reader` program from earlier into mtl-style.
+First, I'll include the original program again:
+
+```tut:book
+def getConfig: IO[Config] = ???
+
+def serviceCall(c: Config): IO[Result] = ???
+
+def readerProgram: ReaderT[IO, Config, Result] = for {
+  config <- ReaderT.ask[IO, Config]
+  result <- ReaderT.liftF(serviceCall(config))
+} yield result
+
+def main: IO[Result] = getConfig.flatMap(readerProgram.run)
+```
+
+Now we should just replace that `ReaderT` with an `F` and add an `ApplicativeAsk[F, Config]` constraint, right?
+We have one small problem though, how can we lift our `serviceCall` which is an `IO` value, into our abstract `F` context?
+Fortunately `cats-effect` already defines a typeclass designed to help us out here called `LiftIO`.
+It defines a single function `liftIO` that does exactly what you'd expect:
+
+```scala
+@typeclass trait LiftIO[F[_]] {
+  def liftIO[A](io: IO[A]): F[A]
+}
+```
+
+If there's an instance for `LiftIO[F]` we can lift any `IO[A]` into an `F[A]`.
+Furthermore `IO` defines a method `to` which makes use of this type class to provide some nicer looking syntax.
+
+With this in mind, we can now define our `readerProgram` fully using MTL:
+
+```tut:silent
+import cats.mtl._
+import cats.mtl.instances.all._
+
+def readerProgram[F[_]: Monad: LiftIO](implicit A: ApplicativeAsk[F, Config]): F[Result] = for {
+  config <- A.ask
+  result <- serviceCall(config).to[F]
+} yield result
+```
+
+We replaced our call to `ReaderT.ask` with a call to `ask` provided by `ApplicativeAsk` and instead of using `ReaderT.liftF` to lift an `IO` into `ReaderT`, we can simply use the `to` function on `IO`, pretty neat if you ask me.
+
+Now to run it, all we need to do is specify the target `F` to run in, in our case `ReaderT[IO, Config, Result]` fits perfectly:
+
+```tut:silent
+val materializedProgram = readerProgram[ReaderT[IO, Config, ?]]
+
+def main: IO[Result] = getConfig.flatMap(materializedProgram.run)
+```
+
+This process of turning a program defined by an abstract type constructor with additional type class constraints into an actual concrete data type is sometimes called *interpreting*  or *materializing*  a program.
+
+Another thing we can do is define a type alias for `ApplicativeAsk[F, Config]` so that we can more easily use it with the context bound syntax:
```

Comment:
Great question, I'm just speculating here, but I think it's because partial unification only works from right to left, so if we have `ApplicativeAsk[E, F[_]]`, it wouldn't be able to unify that with `ReaderT[F[_], E, A]`. We'd have to change the transformers to use that same shape as well. (Which personally I think would be a great idea, but it's kinda of late to do that now)

Reply:
Fair point.